### PR TITLE
Fix and enable FlowRunnerPipelineTest

### PR DESCRIPTION
**Pull request by juhoautio**
Fix and enable FlowRunnerPipelineTest.

Also much less code and faster running.

Separate commits should offer some clarity:
1. first I just fixed the "app initialization" and test file path
1. then I converted the assertions to use the "blocking" versions that wait until the expected status is reached, and removed redundant sleeps
    - Here I also removed the sleep between starting the 1st and 2nd execution: the 2nd execution can be started when the first jobs of the 1st execution have reached RUNNING status. 
1. finally I also converted the flow level assertions to the "blocking" versions

### CODE REVIEW

**Comment by juhoautio** [comment]
Would you like to have these tests? Looks like there's a new conflict to resolve.

**Comment by HappyRay** [comment]
Yes, this change seems to be useful.
Sorry, the team are behind in reviewing PR requests.

File: azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java [hunk 1]
```diff
@@ -67,7 +65,7 @@
  *
  * @author rpark
  */
-public class FlowRunnerPipelineTest {
+public class FlowRunnerPipelineTest extends FlowRunnerTestBase {
```
Comment (chengren311) [inline azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java]: not part of this PR, but I think FlowRunnerTestBase.java needs to be documented better and it's missing license.

File: azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java [hunk 2]
```diff
@@ -499,164 +383,73 @@ public void testBasicPipelineLevel2Run2() throws Exception {
...
-  }
-
-  private void pause(final long millisec) {
-    try {
-      Thread.sleep(millisec);
-    } catch (final InterruptedException e) {
-    }
-  }
-
-  private void createExpectedStateMap(final ExecutableFlowBase flow,
```
Comment (chengren311) [inline azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java]: i'm wondering why was old code doing this check

Comment (juhoautio) [inline azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java]: I really don't see why. After all, `compareStates` only checked for the "expected states" only. Maybe there was a bit more control on something like "checking again that statuses haven't changed on execution 1 although expected them to change on execution 2". But when I was converting these to single line asserts, I didn't encounter any case where it would've seemed important to "check again that statuses haven't changed". So I think it was originally just done in an over-complicated way "by accident" :)

File: azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java (hunk 1 above)
Comment (juhoautio) [inline azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java]: @HappyRay has been wanting to get rid of inheriting FlowRunnerTestBase and having some static utility methods instead.

**Comment by juhoautio** [comment]
Rebased with conflicts resolved.

File: azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java [hunk 3]
```diff
@@ -84,7 +82,7 @@ public FlowRunnerPipelineTest() {
   @Before
   public void setUp() throws Exception {
     System.out.println("Create temp dir");
-    this.workingDir = new File("_AzkabanTestDir_" + System.currentTimeMillis());
+    this.workingDir = new File("build/tmp/_AzkabanTestDir_" + System.currentTimeMillis());
```
Comment (HappyRay) [inline azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java]: Is it better to use junit's temp directory rule?
https://garygregory.wordpress.com/2010/01/20/junit-tip-use-rules-to-manage-temporary-files-and-folders/

File: azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java [hunk 4]
```diff
@@ -97,8 +95,10 @@ public void setUp() throws Exception {
     pluginSet.addPluginClass("test", InteractiveTestJob.class);
     this.fakeExecutorLoader = new MockExecutorLoader();
     this.project = new Project(1, "testProject");
+    Utils.initServiceProvider();
+    JmxJobMBeanManager.getInstance().initialize(new Props());
 
-    final File dir = new File("unit/executions/embedded2");
+    final File dir = new File("../test/execution-test-data/embedded2");
```
Comment (HappyRay) [inline azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerPipelineTest.java]: Please use azkaban.test.executions.ExecutionsTestUtil#getFlowDir instead
See an example at https://github.com/HappyRay/azkaban/blob/1cbe49cce50ffde71784e73c2f03097e83e86780/azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java#L973

**Comment by HappyRay** [comment]
Thanks @chengren311 for reviewing this change.

The new code looks cleaner and it is no longer ignored! Thanks @juhoautio !

**Comment by juhoautio** [comment]
Good stuff @HappyRay 👍 

I have resolve the review comments so far. Anything else?

**Comment by HappyRay** [comment]
Travis CI failure seems to be caused by flaky tests:

:azkaban-solo-server:test FAILED
azkaban.execapp.event.JobCallbackRequestMakerTest > basicPostTest FAILED
    org.junit.runners.model.TestTimedOutException: test timed out after 4000 milliseconds
        at org.apache.log4j.spi.LoggingEvent.<init>(LoggingEvent.java:165)
        at org.apache.log4j.Category.forcedLog(Category.java:391)
        at org.apache.log4j.Category.warn(Category.java:1060)
        at azkaban.execapp.event.JobCallbackRequestMaker.makeHttpRequest(JobCallbackRequestMaker.java:158)
        at azkaban.execapp.event.JobCallbackRequestMakerTest.basicPostTest(JobCallbackRequestMakerTest.java:164)
azkaban.execapp.event.JobCallbackRequestMakerTest > unResponsiveGetTest FAILED
    org.junit.runners.model.TestTimedOutException: test timed out after 4000 milliseconds
        at org.apache.http.client.methods.HttpGet.<init>(HttpGet.java:65)
        at azkaban.execapp.event.JobCallbackUtil.parseJobCallbackProperties(JobCallbackUtil.java:170)
        at azkaban.execapp.event.JobCallbackUtil.parseJobCallbackProperties(JobCallbackUtil.java:94)
        at azkaban.execapp.event.JobCallbackRequestMakerTest.unResponsiveGetTest(JobCallbackRequestMakerTest.java:141)

---
azkaban.execapp.FlowRunnerPipelineTest > testBasicPipelineLevel2Run FAILED
    java.lang.AssertionError: Wrong status for [pipelineEmbeddedFlow3:innerJobB] expected:<RUNNING> but was:<QUEUED>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:118)
        at azkaban.execapp.FlowRunnerTestBase.assertStatus(FlowRunnerTestBase.java:131)
        at azkaban.execapp.FlowRunnerPipelineTest.testBasicPipelineLevel2Run(FlowRunnerPipelineTest.java:315)

**Comment by HappyRay** [comment]
I haven't had time to look into the test logic yet. I've asked other team members to help to review your PRs.

**Comment by juhoautio** [comment]
I went looking for the failed build because your message didn't reveal what the wrong status was :) I found the build: https://travis-ci.org/azkaban/azkaban/builds/266809720

```
azkaban.execapp.FlowRunnerTest2 > testNormalFailure3 FAILED
    java.lang.AssertionError: Wrong status for [jobd] expected:<RUNNING> but was:<READY>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:118)
        at azkaban.execapp.FlowRunnerTestBase.assertStatus(FlowRunnerTestBase.java:117)
        at azkaban.execapp.FlowRunnerTest2.testNormalFailure3(FlowRunnerTest2.java:448)
```

I'll see if I can figure out what could cause it. It's a pity if we can't get the full output of this run from Travis.

**Comment by juhoautio** [comment]
@HappyRay I went looking for the failed build because your message didn't reveal what the wrong status was :) Hmm but I didn't find it. Could you edit your message wrapping the stracktrace into a codeblock?

I'll try to see if I can figure out what could cause it. It's a pity if we can't get the full output of this run from Travis.

**Comment by HappyRay** [comment]
Thanks @juhoautio 

Sorry. I should have included a link. Can you see this page? https://travis-ci.org/azkaban/azkaban/pull_requests if you do, you can search for this PR there.

**Comment by HappyRay** [comment]
I can't find the failed build anymore. I shouldn't have restarted it.

**Comment by juhoautio** [comment]
No worries as I guess there wasn't more to obtain after all. That `expected:<RUNNING> but was:<QUEUED>` was what I was looking for.

**Comment by juhoautio** [comment]
This should be ok as well.

I checked what tests failed in that build c74fad3. They are not related to FlowRunnerPipelineTest. 

```
azkaban.execapp.FlowRunnerTest2 > testPauseFailFinishAll FAILED
    java.lang.IllegalStateException: joba wasn't added in testJobs map
        at azkaban.executor.InteractiveTestJob.getTestJob(InteractiveTestJob.java:55)
        at azkaban.execapp.FlowRunnerTest2.testPauseFailFinishAll(FlowRunnerTest2.java:996)

azkaban.execapp.JobRunnerTest > testDelayedExecutionCancelledJob FAILED
    java.lang.AssertionError
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertTrue(Assert.java:52)
        at azkaban.execapp.JobRunnerTest.testDelayedExecutionCancelledJob(JobRunnerTest.java:303)
:azkaban-web-server:test FAILED

azkaban.execapp.FlowRunnerTest2 > testPause FAILED
    java.lang.AssertionError: Wrong status for [joba] expected:<SUCCEEDED> but was:<RUNNING>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:118)
        at azkaban.execapp.FlowRunnerTestBase.assertStatus(FlowRunnerTestBase.java:131)
        at azkaban.execapp.FlowRunnerTestBase.assertStatus(FlowRunnerTestBase.java:136)
        at azkaban.execapp.FlowRunnerTest2.testPause(FlowRunnerTest2.java:804)
```

**Comment by inramana** [comment]
@juhoautio I was having a look at this over the past week but I am away from now until sep 11. Can this wait or should someone else have a look?

**Comment by juhoautio** [comment]
@inramana, just pinging in case this got buried under other things.. Would be nice to get this in so that we can proceed with the actual bug fix: https://github.com/azkaban/azkaban/pull/1245.

**Comment by HappyRay** [comment]
@juhoautio 

@inramana is busy. I am taking over. The change LGTM. Please resolve the conflict and I will merge it.

**Comment by juhoautio** [comment]
Conflicts resolved.
